data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+use core::{
+    cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd},
+    convert::TryFrom,
+    fmt,
+    hash::{Hash, Hasher},
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+    slice::{from_raw_parts, from_raw_parts_mut},
+};
+#[cfg(feature = "std")]
+use std::str::from_utf8_unchecked;
+
+use derive_more::{DebugCustom, Display};
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::{WrapperTypeDecode, WrapperTypeEncode};
+use serde::{
+    de::{Deserialize, Deserializer, Error, Visitor},
+    ser::{Serialize, Serializer},
+};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (in 64/32/16-bit architecture) are inlined.
+/// Union represents const-string variants: inlined or boxed.
+/// Distinction between variants are achieved by tagging most significant bit of field `len`:
+/// - for inlined variant MSB of `len` is always equal to 1, it's enforced by `InlinedString` constructor;
+/// - for boxed variant MSB of `len` is always equal to 0, it's enforced by the fact
+/// that `Box` and `Vec` never allocate more than`isize::MAX bytes`.
+/// For little-endian 64bit architecture memory layout of `ConstStringData` is following:
+///
+/// ```text
+/// +---------+-------+---------+----------+----------------+
+/// | Bits    | 0..63 | 64..118 | 119..126 | 127            |
+/// +---------+-------+---------+----------+----------------+
+/// | Inlined | payload         | len      | tag (always 1) |
+/// +---------+-------+---------+----------+----------------+
+/// | Box     | ptr   | len                | tag (always 0) |
+/// +---------+-------+--------------------+----------------+
+/// ```
+#[derive(Display, DebugCustom)]
+#[display(fmt = "{}", "&**self")]
+#[debug(fmt = "{:?}", "&**self")]
+#[repr(C)]
+pub union ConstString {
+    inlined: InlinedString,
+    boxed: ManuallyDrop<BoxedString>,
+}
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
+    #[inline]
+    pub fn len(&self) -> usize {
+        if self.is_inlined() {
+            self.inlined().len()
+        } else {
+            self.boxed().len()
+        }
+    }
+
+    /// Returns `true` if [`ConstString`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            inlined: InlinedString::new(),
+        }
+    }
+
+    #[inline]
+    const fn is_inlined(&self) -> bool {
+        // SAFETY: access to the MSB is always safe regardless of the correct variant.
+        self.inlined().is_inlined()
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn inlined(&self) -> &InlinedString {
+        // SAFETY: safe to access if `is_inlined` == `true`.
+        unsafe { &self.inlined }
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn boxed(&self) -> &ManuallyDrop<BoxedString> {
```

Comment:
```suggestion
    const fn boxed(&self) -> &BoxedString {
```

Reply:
ok, but it won't be a `const` function anymore